dilation-protocol.md

```diff
@@ -388,158 +297,53 @@ In addition, the Leader must:
-`counter*2+2` for the Follower, with id `0` reserved for the control
-channel).
-
-When the `client_ep.connect()` API is called, the Initiator allocates a
-subchannel-id and sends an OPEN. It can then immediately send DATA messages
-with the outbound data (there is no special response to an OPEN, so there is
-no need to wait). The Acceptor will trigger their `.connectionMade` handler
-upon receipt of the OPEN.
-
-Subchannels are durable: they do not close until one side calls
-`.loseConnection` on the subchannel object (or the enclosing Wormhole is
-closed). Either the Initiator or the Acceptor can call `.loseConnection`.
-This causes a CLOSE message to be sent (with the subchannel-id). The other
-side will send its own CLOSE message in response. Each side will signal the
-`.connectionLost()` event upon receipt of a CLOSE.
-
-There is no equivalent to TCP's "half-closed" state, however if only one side
-calls `close()`, then all data written before that call will be delivered
-before the other side observes `.connectionLost()`. Any inbound data that was
-queued for delivery before the other side sees the CLOSE will still be
-delivered to the side that called `close()` before it sees
-`.connectionLost()`. Internally, the side which called `.loseConnection` will
-remain in a special "closing" state until the CLOSE response arrives, during
-which time DATA payloads are still delivered. After calling `close()` (or
-receiving CLOSE), any outbound `.write()` calls will trigger an error.
-
-DATA payloads that arrive for a non-open subchannel are logged and discarded.
-
-This protocol calls for one OPEN and two CLOSE messages for each subchannel,
-with some arbitrary number of DATA messages in between. Subchannel-ids should
-not be reused (it would probably work, the protocol hasn't been analyzed
-enough to be sure).
-
-The "control channel" is special. It uses a subchannel-id of all zeros, and
-is opened implicitly by both sides as soon as the first L3 connection is
-selected. It is routed to a special client-on-both-sides endpoint, rather
-than causing the listening endpoint to accept a new connection. This avoids
-the need for application-level code to negotiate who should be the one to
-open it (the Leader/Follower distinction is private to the Wormhole
-internals: applications are not obligated to pick a side).
-
-OPEN and CLOSE messages for the control channel are logged and discarded. The
-control-channel client endpoints can only be used once, and does not close
-until the Wormhole itself is closed.
-
-Each OPEN/DATA/CLOSE message is delivered to the L4 object for queueing,
-delivery, and eventual retirement. The L5 layer does not keep track of old
-messages.
+The DATA payloads are multiplexed over some virtual sub-channels. They may be
+exposed to the application as logically independent data streams.
+
+Each subchannel uses a distinct subchannel-id, which is a big-endian four-byte identifier.
```

Comment:
Should include some mention that subchannel-ids are a "protocol" thing and *should not* be leaked through the application API.

(That is, application shouldn't be allowed to select or depend upon subchannel IDs).

Reply:
Shouldn't it? I mean, I don't have any strong opinions on this, but the initial protocol said nothing about that so I'd like to think it through some more. Are there scenarios where applications might want to reference some communication channel for some reason? Does it hurt us to expose the channel number?

(Of course we can always add the restriction now to be sure and then lift it later on should there be need …)

Reply:
Yes, that's what Brian and I talked about some time back. I'm not sure if he wrote this down anywhere?

To me it does make sense, now; there are requirements on the numbers so they should be picked by the protocol. If an application does need to cross-reference channels somehow I think it makes way more sense for that application-protocol to add messages that make sense to that application. (For example, they could have an initial message indicating their name or ID or whatever).